# Code 11 to Code 16

In [1]:
import pandas as pd
import re


# Tokens : Are individual units of analysis
# simplest way to get tokens is to use 'split()'
# documents : Are individual records

In [2]:
str1 = "What a great show!"

In [3]:
str1.split()

['What', 'a', 'great', 'show!']

As seen above the simplest way to break sentences to words is to use “split” function. However split does not accomodate special characters whereas “word_tokenize” from nltk does take care of special characters.Sentences can be split by using “sent_tokenize” function as shown below

In [4]:
from nltk.tokenize import word_tokenize

In [5]:
word_tokenize(str1)

['What', 'a', 'great', 'show', '!']

For Tokenizing sentences

In [6]:
from nltk.tokenize import sent_tokenize

In [7]:
sentences = "I am feeling great! It was a great show"

In [8]:
sentences.split("!")

['I am feeling great', ' It was a great show']

In [9]:
from nltk.tokenize import sent_tokenize
sent_tokenize(sentences)

['I am feeling great!', 'It was a great show']

Sometimes we may also want to analyze words in a certain order. For example “check in” means something together and the value of that is lost if it analyzed as “check” and “in”. So in order to retain as much meaning as possible while converting sentences to word tokens we also want to retain the order. So instead of having single word tokens we could 2 or 3 term tokens. They are referred as bigrams and trigrams. 

In [10]:
import nltk
text = "I want to check in asap"
bigram_list = list(nltk.bigrams(text.split()))
[' '.join(i) for i in bigram_list]

['I want', 'want to', 'to check', 'check in', 'in asap']

In [11]:
import numpy as np

Let us import a datset and apply tokeization

In [12]:
def combine(list1):
    str_all = ' '.join(list(ser))
    
    return str_all

In [13]:
def tokenize(sent1):
    return word_tokenize(sent1)

In [14]:
df = pd.read_csv('airline_dataset.csv',low_memory=False,encoding = 'latin1')
df["line"]=df["line"].str.lower().str.lstrip().str.rstrip()

In [15]:
###tokenize all rows
df["token_words"] = df["line"].apply(tokenize)

Bag of words from each of the sentence

In [16]:
df.head()

line     class  \
0          when can i web check-in?  check in   
1                  want to check in  check in   
2                please check me in  check in   
3                          check in  check in   
4  my flight is tomm can i check in  check in   

                                 token_words  
0           [when, can, i, web, check-in, ?]  
1                      [want, to, check, in]  
2                    [please, check, me, in]  
3                                [check, in]  
4  [my, flight, is, tomm, can, i, check, in]

In [17]:
df.head().to_clipboard()

get unique list of words from all sentences

In [18]:

list_all = []
for i in df["token_words"]:
    list_all = list_all + i
list_words = list(set(list_all))
list_words[0:5]

['login', 'insurance', 'kg', 'am', 'a']

initialize array for rows=length of dataframe and columns = unique words.

In [19]:

arr1 = np.zeros([len(df),len(list_words)])

Loop through the dataframe and mark presence of words in a sentence as 1

In [20]:

for i,row in df.iterrows():
    token_words = row["token_words"]
    for j in token_words:
        if (j in list_words):
            k = list_words.index(j)
            arr1[i][k] = 1

In [21]:
###Check if we got it right
get_non_zero = np.where(arr1[0]==1)
list_index = list(get_non_zero[0])
[list_words[i] for i in list_index]

['i', '?', 'when', 'can', 'web', 'check-in']

In [22]:
df_mat = pd.DataFrame(arr1)
df_mat.columns = list_words

In [23]:
df_mat.head()

login  insurance   kg   am    a  aircraft  seats  much   be  abel  ...  \
0    0.0        0.0  0.0  0.0  0.0       0.0    0.0   0.0  0.0   0.0  ...   
1    0.0        0.0  0.0  0.0  0.0       0.0    0.0   0.0  0.0   0.0  ...   
2    0.0        0.0  0.0  0.0  0.0       0.0    0.0   0.0  0.0   0.0  ...   
3    0.0        0.0  0.0  0.0  0.0       0.0    0.0   0.0  0.0   0.0  ...   
4    0.0        0.0  0.0  0.0  0.0       0.0    0.0   0.0  0.0   0.0  ...   

   sir  mmmm  have  nice  with  return  next  checking  money  failed  
0  0.0   0.0   0.0   0.0   0.0     0.0   0.0       0.0    0.0     0.0  
1  0.0   0.0   0.0   0.0   0.0     0.0   0.0       0.0    0.0     0.0  
2  0.0   0.0   0.0   0.0   0.0     0.0   0.0       0.0    0.0     0.0  
3  0.0   0.0   0.0   0.0   0.0     0.0   0.0       0.0    0.0     0.0  
4  0.0   0.0   0.0   0.0   0.0     0.0   0.0       0.0    0.0     0.0  

[5 rows x 232 columns]

Above is a quick visualization of the matrix we created. This is called the term-document matrix. Remember we filled the cells with 0 or 1 representing presence or absence of a word in a sentence. We could also mark the cells as “frequency” of words in the sentence. Also in the above example we see we generated 232 features for a corpus size of 398 documents. Hence we see a plenty of zeros in the above matrix